File: scala/src/main/scala/datastructures/MinHeap.scala
package datastructures

import scala.collection.mutable.ArrayBuffer

/**
 * Min-heap implementation with custom comparator support.
 * Time: O(log n) for insert/extract, O(1) for peek
 * Space: O(n)
 */
class MinHeap[T](using ord: Ordering[T]):
  private val items = ArrayBuffer.empty[T]

  def insert(value: T): this.type =
    items += value
    siftUp(items.length - 1)
    this

  def extractMin(): Option[T] =
    if items.isEmpty then None
    else
      val min = items(0)
      val last = items.remove(items.length - 1)
      if items.nonEmpty then
        items(0) = last
        siftDown(0)
      Some(min)

  def peek: Option[T] =
    if items.isEmpty then None
    else Some(items(0))

  def isEmpty: Boolean = items.isEmpty

  def size: Int = items.length

  def clear(): this.type =
    items.clear()
    this

  def toList: List[T] = items.toList

  private def siftUp(index: Int): Unit =
    var i = index
    while i > 0 do
      val parent = (i - 1) / 2
      if ord.lt(items(i), items(parent)) then
        swap(i, parent)
        i = parent
      else
        return

  private def siftDown(index: Int): Unit =
    var i = index
    while true do
      val left = 2 * i + 1
      val right = 2 * i + 2
      var smallest = i

      if left < items.length && ord.lt(items(left), items(smallest)) then
        smallest = left
      if right < items.length && ord.lt(items(right), items(smallest)) then
        smallest = right

      if smallest != i then
        swap(i, smallest)
        i = smallest
      else
        return

  private def swap(i: Int, j: Int): Unit =
    val temp = items(i)
    items(i) = items(j)
    items(j) = temp

object MinHeap:
  def from[T: Ordering](elements: Iterable[T]): MinHeap[T] =
    val heap = new MinHeap[T]
    for elem <- elements do heap.insert(elem)
    heap

File: haskell/src/DataStructures/MinHeap.hs
{-|
Module      : DataStructures.MinHeap
Description : Min-heap (priority queue) implementation
Stability   : stable

A purely functional min-heap implementation.
Time: O(log n) for insert/extract, O(1) for peek
Space: O(n)
-}
module DataStructures.MinHeap
  ( MinHeap
  , empty
  , isEmpty
  , insert
  , extractMin
  , peek
  , size
  , fromList
  , toList
  ) where

-- | A min-heap represented as a leftist heap
data MinHeap a = Empty | Node Int a (MinHeap a) (MinHeap a)
  deriving (Show, Eq)

-- | Get rank of heap (length of right spine)
rank :: MinHeap a -> Int
rank Empty          = 0
rank (Node r _ _ _) = r

-- | Create an empty heap
empty :: MinHeap a
empty = Empty

-- | Check if heap is empty
isEmpty :: MinHeap a -> Bool
isEmpty Empty = True
isEmpty _     = False

-- | Make a node with correct rank (leftist property)
makeNode :: a -> MinHeap a -> MinHeap a -> MinHeap a
makeNode x left right
  | rank left >= rank right = Node (rank right + 1) x left right
  | otherwise               = Node (rank left + 1) x right left

-- | Merge two heaps
merge :: Ord a => MinHeap a -> MinHeap a -> MinHeap a
merge Empty h = h
merge h Empty = h
merge h1@(Node _ x1 l1 r1) h2@(Node _ x2 l2 r2)
  | x1 <= x2  = makeNode x1 l1 (merge r1 h2)
  | otherwise = makeNode x2 l2 (merge h1 r2)

-- | Insert a value
insert :: Ord a => a -> MinHeap a -> MinHeap a
insert x h = merge (Node 1 x Empty Empty) h

-- | Extract the minimum value
extractMin :: Ord a => MinHeap a -> Maybe (a, MinHeap a)
extractMin Empty          = Nothing
extractMin (Node _ x l r) = Just (x, merge l r)

-- | Peek at minimum without removing
peek :: MinHeap a -> Maybe a
peek Empty          = Nothing
peek (Node _ x _ _) = Just x

-- | Get number of elements
size :: MinHeap a -> Int
size Empty            = 0
size (Node _ _ l r) = 1 + size l + size r

-- | Create heap from list
fromList :: Ord a => [a] -> MinHeap a
fromList = foldr insert empty

-- | Convert to sorted list
toList :: Ord a => MinHeap a -> [a]
toList h = case extractMin h of
  Nothing      -> []
  Just (x, h') -> x : toList h'
